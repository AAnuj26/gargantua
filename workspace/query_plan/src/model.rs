use std::ops::{Deref, DerefMut};

use async_graphql::Positioned;
use async_graphql_parser::types::{self as Q};
use blueprint::{Graph, JoinField};
use derive_setters::Setters;

use crate::error::Error;

#[derive(Debug, Clone)]
pub enum QueryPlan<Value> {
    Parallel(Vec<QueryPlan<Value>>),
    Sequence(Vec<QueryPlan<Value>>),
    Fetch {
        service: Graph,
        query: QueryOperation<Value>,
        representations: Option<SelectionSet<Value>>,
        type_name: TypeName,
    },
    Flatten {
        select: Lens,
        plan: Box<QueryPlan<Value>>,
    },
}

#[derive(Debug, Clone)]
pub struct QueryOperation<Value> {
    // TODO: add directives, variables etc.
    pub selection_set: SelectionSet<Value>,
}

impl<A> QueryPlan<A> {
    pub fn fetch(service: Graph, type_name: TypeName, query: SelectionSet<A>) -> Self {
        QueryPlan::Fetch {
            service,
            query: QueryOperation { selection_set: query },
            representations: None,
            type_name,
        }
    }
}

impl QueryPlan<async_graphql_value::Value> {
    // Tries to create a new Query Plan from a GraphQL query and a Blueprint Index.
    pub fn try_new(query: &str) -> Result<Self, Error> {
        let doc = async_graphql_parser::parse_query(query)?;
        let mut parallel = Vec::new();

        // TODO: handle fragments
        // TODO: use named operations
        for (_, op) in doc.operations.iter() {
            let selection = SelectionSet::from(&op.node.selection_set.node);
            let type_name = TypeName::new("Query");
            let service = Graph::new("WIP");
            parallel.push(QueryPlan::fetch(service, type_name, selection));
        }
        Ok(QueryPlan::Parallel(parallel))
    }

    // Sequentially executes one plan after the other
    pub fn and_then(self, select: Lens, plan: QueryPlan<async_graphql_value::Value>) -> Self {
        QueryPlan::Sequence(vec![
            self,
            QueryPlan::Flatten { select, plan: Box::new(plan) },
        ])
    }
}

#[derive(Debug, Clone)]
pub struct TypeName(String);

impl TypeName {
    pub fn new(name: &str) -> Self {
        TypeName(name.to_string())
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

#[derive(Default, Debug, Clone)]
pub struct SelectionSet<Value>(Vec<Field<Value>>);

impl<A> Deref for SelectionSet<A> {
    type Target = Vec<Field<A>>;

    fn deref(&self) -> &Self::Target {
        self.0.as_ref()
    }
}

impl<Value> SelectionSet<Value> {
    pub fn new(fields: Vec<Field<Value>>) -> Self {
        Self(fields)
    }

    pub fn push(&mut self, field: Field<Value>) {
        self.0.push(field);
    }

    pub fn into_vec(self) -> Vec<Field<Value>> {
        self.0
    }
}

#[derive(Debug, Clone, Setters)]
pub struct Field<Value> {
    pub name: String,
    pub selections: SelectionSet<Value>,
    pub arguments: Vec<Argument<Value>>,
    pub directives: Vec<Directive<Value>>,

    /// When set to true the field is considered to be used internally for
    /// querying sub-graphs and should not be exposed to the user.
    pub is_hidden: bool,

    /// Possible Graphs from where the field can be queried from.
    pub graph: Vec<Graph>,

    /// Internal readonly information from the Blueprint Index.
    // #[setters(skip)]
    join_field: Vec<JoinField>,
}

impl<A> Field<A> {
    // pub join_field(&mut)

    pub fn new(name: String, selections: SelectionSet<A>) -> Self {
        Field {
            name: name.to_string(),
            selections,
            arguments: Vec::new(),
            directives: Vec::new(),
            is_hidden: false,
            graph: Vec::new(),
            join_field: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, Setters)]
pub struct Argument<Value> {
    pub name: String,
    pub value: Value,
}

#[derive(Debug, Clone, Setters)]
pub struct Directive<Value> {
    pub name: String,
    pub arguments: Vec<Argument<Value>>,
}

#[derive(Debug, Clone)]
pub enum Lens {
    Field(String),
    Index(usize),
    Combine(Box<Lens>, Box<Lens>),
    ForEach(Box<Lens>),
    Empty,
}

impl Lens {
    pub fn get(&self, _value: serde_json::Value) -> serde_json::Value {
        // TODO: implement
        todo!()
    }
    pub fn set(
        &self,
        _value: serde_json::Value,
        _other_value: serde_json::Value,
    ) -> serde_json::Value {
        // TODO: implement
        todo!()
    }
}

// Correctly implement and add tests
impl From<&Q::SelectionSet> for SelectionSet<async_graphql_value::Value> {
    fn from(node: &Q::SelectionSet) -> SelectionSet<async_graphql_value::Value> {
        let mut selection_set = Vec::new();
        for selection in node.items.iter() {
            let inner_selection = &selection.node;
            match inner_selection {
                Q::Selection::Field(Positioned { node, .. }) => {
                    let field_name = node.name.node.as_str().to_string();
                    let field =
                        Field::new(field_name, SelectionSet::from(&node.selection_set.node));
                    selection_set.push(field);
                }
                Q::Selection::InlineFragment(_) => {
                    todo!()
                }
                Q::Selection::FragmentSpread(_) => {
                    todo!()
                }
            }
        }
        SelectionSet(selection_set)
    }
}

#[cfg(test)]
mod test {
    use insta::assert_debug_snapshot;

    use crate::QueryPlan;

    #[test]
    fn test() {
        let query = "query { topProducts { name reviews { score } reviews { description } } }";
        let actual: QueryPlan<_> = QueryPlan::try_new(query).unwrap();
        assert_debug_snapshot!(actual);
    }

    #[test]
    fn test_complex() {
        let query = r#"
            query getData(
                $userId: String!
                $sortOrder: String = DESC
                $region: String = "EU"
            ) @onQuery {
                me: user(id: $userId) @onField {
                    id
                    username
                    role {
                    id
                    name
                    }
                }
                stores(first: 10, order: $sortOrder, region: $region) {
                    id @onField(data: 1)
                    name @onField(data: { foo: "bar" })
                }
            }
        "#;
        let actual: QueryPlan<_> = QueryPlan::try_new(query).unwrap();
        assert_debug_snapshot!(actual);
    }
}
