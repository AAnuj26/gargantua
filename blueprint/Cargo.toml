[package]
name = "nova"
version = "0.1.0"
edition = "2021"

[dependencies]
derive_setters.workspace = true
serde.workspace = true
serde_json = "1.0.128"
