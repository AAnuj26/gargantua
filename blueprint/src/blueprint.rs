use std::collections::{BTreeMap, BTreeSet};

use async_graphql_parser::types::ServiceDocument;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use valid::Valid;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Blueprint {
    pub definitions: Vec<Definition>,
    pub schema: SchemaDefinition,
    pub directives: Vec<DirectiveDefinition>,
    pub graphs: Vec<JoinGraph>,
}

#[derive(Clone, Debug, Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
pub struct GraphId(pub String);

#[derive(Clone, Debug, Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
pub struct JoinGraph {
    pub name: GraphId,
    pub url: url::Url,
}

impl Blueprint {
    pub fn parse(doc: ServiceDocument) -> Valid<Blueprint, String> {
        super::parse::parse(doc)
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum Definition {
    Interface(InterfaceTypeDefinition),
    Object(ObjectTypeDefinition),
    InputObject(InputObjectTypeDefinition),
    Scalar(ScalarTypeDefinition),
    Enum(EnumTypeDefinition),
    Union(UnionTypeDefinition),
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InterfaceTypeDefinition {
    pub name: String,
    pub fields: Vec<FieldDefinition>,
    pub description: Option<String>,
    pub join_type: Vec<JoinType>,
    pub join_implements: Vec<JoinImplements>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ObjectTypeDefinition {
    pub name: String,
    pub fields: Vec<FieldDefinition>,
    pub description: Option<String>,
    pub implements: BTreeSet<String>,
    pub join_type: Vec<JoinType>,
    pub join_implements: Vec<JoinImplements>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InputObjectTypeDefinition {
    pub name: String,
    pub fields: Vec<InputFieldDefinition>,
    pub description: Option<String>,
    pub join_type: Vec<JoinType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct EnumTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub enum_values: Vec<EnumValueDefinition>,
    pub join_type: Vec<JoinType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct EnumValueDefinition {
    pub description: Option<String>,
    pub name: String,
    pub directives: Vec<Directive>,
    pub join_enum: Vec<GraphId>
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct SchemaDefinition {
    pub query: Option<String>,
    pub mutation: Option<String>,
    pub subscription: Option<String>,
    pub directives: Vec<Directive>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InputFieldDefinition {
    pub name: String,
    pub of_type: Type,
    pub default_value: Option<Value>,
    pub description: Option<String>,
    pub join_field: Vec<JoinField>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct FieldDefinition {
    pub name: String,
    pub args: Vec<InputFieldDefinition>,
    pub of_type: Type,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub join_field: Vec<JoinField>
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Directive {
    pub name: String,
    pub arguments: BTreeMap<String, Value>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DirectiveDefinition {
    pub name: String,
    pub description: Option<String>,
    pub arguments: Vec<InputFieldDefinition>,
    pub repeatable: bool,
    pub locations: Vec<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ScalarTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub join_type: Vec<JoinType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UnionTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub types: BTreeSet<String>,
    pub join_type: Vec<JoinType>,
    pub join_union: Vec<JoinUnion>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinType {
    pub graph: GraphId,
    pub key: Option<String>,
    pub extension: bool,
    pub resolvable: bool,
    pub is_interface_object: bool,
}

impl JoinType {
    pub fn new(graph: String) -> Self {
        Self {
            graph: GraphId(graph),
            key: None,
            extension: false,
            resolvable: true,
            is_interface_object: false,
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinField {
    pub graph: Option<GraphId>,
    pub requires: Option<String>,
    pub provides: Option<String>,
    pub r#type: Option<String>,
    pub external: Option<bool>,
    pub r#override: Option<String>,
    pub used_overridden: Option<bool>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinImplements {
    pub graph: GraphId,
    pub interface: String
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinUnion {
    pub graph: GraphId,
    pub member: String
}

/// Type to represent GraphQL type usage with modifiers
/// [spec](https://spec.graphql.org/October2021/#sec-Wrapping-Types)
#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum Type {
    Named {
        /// Name of the type
        name: String,
        /// Flag to indicate the type is required.
        required: bool,
    },
    List {
        /// Type is a list
        of_type: Box<Type>,
        /// Flag to indicate the type is required.
        non_null: bool,
    },
}
