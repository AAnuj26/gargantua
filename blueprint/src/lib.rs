mod blueprint;
mod parse;
pub use blueprint::*;

pub use url;