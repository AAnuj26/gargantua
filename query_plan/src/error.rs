use derive_more::From;
use valid::ValidationError;

#[derive(From)]
pub enum Error {
    Blueprint(blueprint::error::Error),
    Parse(async_graphql_parser::Error),

    // Error while creating the query plan
    Plan(ValidationError<String>),
}
