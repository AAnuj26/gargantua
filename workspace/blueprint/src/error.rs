use derive_more::From;

#[derive(Debug, From)]
pub enum Error {
    Parse(async_graphql_parser::Error),
}
