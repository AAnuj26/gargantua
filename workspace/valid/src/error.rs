use std::fmt::{Debug, Display};

use super::Cause;

#[derive(Debug, PartialEq, Default, Clone)]
pub struct Error<E>(Vec<Cause<E>>);

impl<E: Display> Display for Error<E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("Validation Error\n")?;
        let errors = self.as_vec();
        for error in errors {
            f.write_str(format!("{} {}", '\u{2022}', error.message).as_str())?;
            if !error.trace.is_empty() {
                f.write_str(
                    &(format!(
                        " [{}]",
                        error
                            .trace
                            .iter()
                            .cloned()
                            .collect::<Vec<String>>()
                            .join(", ")
                    )),
                )?;
            }
            f.write_str("\n")?;
        }

        Ok(())
    }
}

impl<E> Error<E> {
    pub fn as_vec(&self) -> &Vec<Cause<E>> {
        &self.0
    }

    pub fn combine(mut self, mut other: Error<E>) -> Error<E> {
        self.0.append(&mut other.0);
        self
    }

    pub fn empty() -> Self {
        Error(Vec::new())
    }

    pub fn new(e: E) -> Self {
        Error(vec![Cause::new(e)])
    }

    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }

    pub fn trace(self, message: &str) -> Self {
        let mut errors = self.0;
        for cause in errors.iter_mut() {
            cause.trace.insert(0, message.to_owned());
        }
        Self(errors)
    }

    pub fn append(self, error: E) -> Self {
        let mut errors = self.0;
        errors.push(Cause::new(error));
        Self(errors)
    }

    pub fn transform<E1>(self, f: &impl Fn(E) -> E1) -> Error<E1> {
        Error(self.0.into_iter().map(|cause| cause.transform(f)).collect())
    }
}

impl<E: Display + Debug> std::error::Error for Error<E> {}

impl<E> From<Cause<E>> for Error<E> {
    fn from(value: Cause<E>) -> Self {
        Error(vec![value])
    }
}

impl<E> From<Vec<Cause<E>>> for Error<E> {
    fn from(value: Vec<Cause<E>>) -> Self {
        Error(value)
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;
    use stripmargin::StripMargin;

    use crate::{Cause, Error};

    #[derive(Debug, PartialEq, serde::Deserialize)]
    struct Foo {
        a: i32,
    }

    #[test]
    fn test_error_display_formatting() {
        let error = Error::from(vec![
            Cause::new("1").trace(vec!["a", "b"]),
            Cause::new("2"),
            Cause::new("3"),
        ]);
        let expected_output = "\
        |Validation Error
        |• 1 [a, b]
        |• 2
        |• 3
        |"
        .strip_margin();
        assert_eq!(format!("{}", error), expected_output);
    }
}
