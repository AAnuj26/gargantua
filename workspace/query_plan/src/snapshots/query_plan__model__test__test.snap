---
source: workspace/query_plan/src/model.rs
expression: actual
---
Parallel(
    [
        Fetch {
            service: Graph(
                "WIP",
            ),
            query: QueryOperation {
                selection_set: SelectionSet(
                    [
                        Field {
                            name: "topProducts",
                            selections: SelectionSet(
                                [
                                    Field {
                                        name: "name",
                                        selections: SelectionSet(
                                            [],
                                        ),
                                        arguments: [],
                                        directives: [],
                                        is_hidden: false,
                                        graph: [],
                                        join_field: [],
                                    },
                                    Field {
                                        name: "reviews",
                                        selections: SelectionSet(
                                            [
                                                Field {
                                                    name: "score",
                                                    selections: SelectionSet(
                                                        [],
                                                    ),
                                                    arguments: [],
                                                    directives: [],
                                                    is_hidden: false,
                                                    graph: [],
                                                    join_field: [],
                                                },
                                            ],
                                        ),
                                        arguments: [],
                                        directives: [],
                                        is_hidden: false,
                                        graph: [],
                                        join_field: [],
                                    },
                                    Field {
                                        name: "reviews",
                                        selections: SelectionSet(
                                            [
                                                Field {
                                                    name: "description",
                                                    selections: SelectionSet(
                                                        [],
                                                    ),
                                                    arguments: [],
                                                    directives: [],
                                                    is_hidden: false,
                                                    graph: [],
                                                    join_field: [],
                                                },
                                            ],
                                        ),
                                        arguments: [],
                                        directives: [],
                                        is_hidden: false,
                                        graph: [],
                                        join_field: [],
                                    },
                                ],
                            ),
                            arguments: [],
                            directives: [],
                            is_hidden: false,
                            graph: [],
                            join_field: [],
                        },
                    ],
                ),
            },
            representations: None,
            type_name: TypeName(
                "Query",
            ),
        },
    ],
)
